funnel/models/project_membership.py

```diff
@@ -0,0 +1,187 @@
+            },
+            'call': {'url_for'},
+        },
+        'editor': {'read': {'edit_url', 'delete_url'}},
+    }
+
+    project_id = immutable(
+        db.Column(None, db.ForeignKey('project.id', ondelete='CASCADE'), nullable=False)
+    )
+    project = immutable(
+        db.relationship(
+            Project,
+            backref=db.backref(
+                'crew_memberships',
+                lazy='dynamic',
+                cascade='all, delete-orphan',
+                passive_deletes=True,
+            ),
+        )
+    )
+    parent = immutable(db.synonym('project'))
+
+    user_id = immutable(
+        db.Column(
+            None,
+            db.ForeignKey('user.id', ondelete='CASCADE'),
+            nullable=False,
+            index=True,
+        )
+    )
+    user = with_roles(
+        immutable(
+            db.relationship(
+                User,
+                foreign_keys=[user_id],
+                backref=db.backref(
+                    'profile_crew_memberships',
+                    lazy='dynamic',
+                    cascade='all, delete-orphan',
+                    passive_deletes=True,
+                ),
+            )
+        ),
+        grants={'subject'},
+    )
+
+    # Project crew roles (at least one must be True):
+
+    #: Editors can edit all common and editorial details of an event
+    is_editor = db.Column(db.Boolean, nullable=False, default=False)
+    #: Concierges are responsible for logistics and have write access
+    #: to common details plus read access to everything else. Unlike
+    #: editors, they cannot edit the schedule
+    is_concierge = db.Column(db.Boolean, nullable=False, default=False)
+    #: Ushers help participants find their way around an event and have
+    #: the ability to scan badges at the door
+    is_usher = db.Column(db.Boolean, nullable=False, default=False)
+
+    @declared_attr
+    def __table_args__(cls):
+        args = list(super(cls, cls).__table_args__)
+        args.append(
+            db.CheckConstraint(
+                'is_editor IS TRUE OR is_concierge IS TRUE OR is_usher IS TRUE',
+                name='project_crew_membership_has_role',
+            )
+        )
+        return tuple(args)
+
+    @property
+    def user_details(self):
+        return {
+            'fullname': self.user.fullname,
+            'username': self.user.username,
+            'avatar': self.user.avatar,
+        }
+
+    @property
+    def edit_url(self):
+        return self.url_for('edit', _external=True)
+
+    @property
+    def delete_url(self):
+        return self.url_for('delete', _external=True)
+
+    def offered_roles(self):
+        """Roles offered by this membership record"""
+        roles = set()
+        if self.is_editor:
+            roles.add('project_editor')
+        if self.is_concierge:
+            roles.add('project_concierge')
+        if self.is_usher:
+            roles.add('project_usher')
+        return roles
```

Comment:
Unlike with the `profile_owner` thing above, here the `editor` and `project_editor` roles need to be kept distinct, as `editor` means anyone who is authorised to edit the project, which includes the concierge role.

For the sake of consistency we should also retain the `profile_*` prefixes.